### PR TITLE
AUDIO: [WIP, RFC] Sync with upstreams Nuked OPL. Last commit: 730f8c2

### PR DESCRIPTION
**Pull request by eriktorbjorn**
Prompted by https://bugs.scummvm.org/ticket/14770

This is my attempt to sync ScummVM's Nuked OPL with upstreams. However, this brings in the C99 integer types (int16_t, and the likes). ScummVM provides those through common/inttypes.h but not always the following:

* uint64_t
* uint_fast32_t
* The UINT64_C() macro

Something may have to be done about that. (In fact, I think we used to have our own solution to the UINT64_C() thing, but that got removed here. Perhaps it should be put back?)

Also, the bulk of the changes appear to be two optional features: OPL_QUIRK_CHANNELSAMPLEDELAY and OPL_ENABLE_STEREOEXT. I have no idea if ScummVM could/should enable those.

### CODE REVIEW

**Comment by ccawley2011** [comment]
Does ScummVM still support compiling with MSVC 2012 or earlier? If we don't, it should be possible to include inttypes.h directly in common/scummsys.h and remove the compatibility typedefs.

**Comment by bluegr** [comment]
> Does ScummVM still support compiling with MSVC 2012 or earlier? If we don't, it should be possible to include inttypes.h directly in common/scummsys.h and remove the compatibility typedefs.

No, the earliest VS version supported is 2015

**Comment by sev-** [comment]
No, we should not include system-wide includes not in backend. e.g. `#include <inttypes.h>` is a no-no for portability reasons.

**Comment by sev-** [comment]
I saw these changes in the upstream, and the only change that I saw could be applicable is fixing UB. Another functional change is 4 channel output which I believe, us not used by us.

**Comment by sev-** [comment]
I understand, but prefer not to open this can of worms where we think, which of the system-wide includes are OK and which are not, and why.

**Comment by bluegr** [comment]
These types are not standardized unfortunately, but we could define them in our portdefs.h, to avoid opening this can of worms, as @sev- mentioned. [Boost checks for the availability of `long long`](https://github.com/boostorg/config/blob/master/include/boost/cstdint.hpp), but we don't, as we use it as a type in several places in our codebase.

In particular:
- `uint64_t` is not really standardized, although, if I'm not mistaken, we could do this, for our needs:
    `typedef unsigned long long uint64_t;`
    Check [here](https://stackoverflow.com/questions/32198368/unsigned-long-long-conflict-with-uint64-t)
- `uint_fast32_t` differs from `uint32_t`, as it's a type that has *at least 32 bits*, instead of *exactly 32 bits*. For our needs, we could use the same typedef as `uint32_t`, I suppose?
  Check [here](https://stackoverflow.com/questions/8500677/what-is-uint-fast32-t-and-why-should-it-be-used-instead-of-the-regular-int-and-u)
- As for the UINT64_C() macro, boost defines it like this, perhaps we could do the same?
  `#  define UINT64_C(value)   value##ui64`
  Check [here](https://stackoverflow.com/questions/52490211/purpose-of-using-uint64-c)

So, to sum up, we could add the following in our portdefs.h, or we can just add these in nuked.h:

    typedef unsigned long long uint64_t;
    typedef unsigned long uint_fast32_t;
    #  define UINT64_C(value)   value##ui64

**Comment by lephilousophe** [comment]
I think these should go in nuked.h instead of our common includes.
Some engines also define these types so that would collide.
I also propose to reuse our own types instead:
```
typedef uint64 uint64_t;
typedef uint32 uint_fast32_t;
#  define UINT64_C(value)   value##ui64
```
We may want to use uintptr instead of uint32 for the fast version as operations may be faster but I am not sure about this.

**Comment by bluegr** [comment]
Implemented the type defines, as discussed here in PR #5895 

Closing this PR in favor of the other one
